README.md
# MiniTemplator ver 1.2

MiniTemplator is a java compact template engine.

# Features

Simple syntax for template variables and blocks.
Blocks can be nested.
Compact, well-documented source code. A single code module.
Documentation and sample programs available.
Open Source, licensed under the [GNU LGPL](http://www.gnu.org/licenses/lgpl.html) & [EPL](http://www.eclipse.org/legal). It may be used freely, even for commercial projects ([Details](http://en.wikipedia.org/wiki/LGPL)).
Available as maven dependency from [CENTRAL](http://search.maven.org/#artifactdetails%7Corg.bsc.util%7Cminitemplator-repackaged%7C1.2%7Cmaven-plugin)

```xml
<dependency>
    <groupId>org.bsc.util</groupId>
    <artifactId>minitemplator-repackaged</artifactId>
    <version>1.1</version>
</dependency>
```

# What is a Template Engine?

A template engine is a software module which is used to generate HTML pages (or other text files). The layout of a HTML page is defined within a template file, which can be designed by using a standard HTML editor like FrontPage. At run-time, the application program instructs the template engine to load the template, fill in variable values, repeat blocks and produce the final HTML page.

# Why use a Template Engine?

The advantage of using a template engine, instead of embedding program code and HTML code, is:

 * Program code and web page layout are separated and can be designed and modified independently.
 * A software developer and a web page designer can work together without interfering in each others work. Responsibilities can be split.
 * After the program has been finished, the page layout can be modified at a later time, without having to modify the program code.
 * Templates are useful when web pages will be translated, and multiple languages must be supported by the same program.

# Why use MiniTemplator?

 * There are other template engines (like Velocity, WebMacro or Smarty) which are huge and use complex template syntaxes. These template syntaxes are so complex that they are almost like new programming languages. MiniTemplator however uses a very simple template syntax. All processing (loops, if-then-else logic) and variable formatting is done in application program code (Java/PHP/VB), which leads to better separation of layout and code, and no new complex syntax has to be learned.
 * There are simple template engines which do not provide elementary features (like nested blocks, loading the template from a string, or storing the output into a string variable), or are slow. MiniTemplator provides these elementary features and is fast, because the template is only parsed once (at the time it is loaded).
 * MiniTemplator provides the features to build professional web applications, and it is compact and fits within a single code module.

## Syntax

A simple syntax is used to define variables and blocks within the template. This syntax is compatible with HTML editors like FrontPage, but it can also be used for non-HTML files.

## Variables:

```
${VariableName}
```

## Escape Variables (ignore)

```
\${VariableName}
```
result will be
```
${VariableName}
```
## Blocks:

```
<!-- $BeginBlock BlockName -->
 ... block content ...
<!-- $EndBlock BlockName -->
```

## Include a subtemplate:

```
<!-- $Include RelativeFileName -->
```


CHANGES.txt

MiniTemplator Changes (Java version)
------------------------------------

see 
commits: https://github.com/bsorrentino/minitemplator/commits/master
issues: https://github.com/bsorrentino/minitemplator/issues

2011-11-27 chdh
Short form syntax added for conditional blocks.

2011-05-17 chdh
Directory structure changed to make it Maven compliant.

2010-10-30 chdh
Error corrected: When the conditionFlags set was empty or null, the
block of an IF statement in the form "$if !flag" was not included
in the output.

2010-09-29 chdh
Convenience methods MiniTemplator.setVariable(String, int) and
setVariableOpt(String, int) added.

2010-02-19 chdh
Protected method MiniTemplator.newInstance() added to enable
subclasses of MiniTemplator to survive cloneReset().

2009-04-15 chdh
The complement ("not") operator "!" may now be used in the flag
expressions of the $if and $elseif commands.

2009-01-25 chdh
New constructor MiniTemplator(String) added.
Note that this constructor is not compatible with the old (prior to
2007-05-19) constructor with the same signature, because the old
constructor with that signature expected the template string in the
string argument instead of the file name.
The nested exceptions (TemplateSyntaxException,
VariableNotDefinedException and BlockNotDefinedException) are now
derived from RuntimeException instead of Exception, to make them
unchecked exceptions that do not have to be catched or declared.
Convenience methods added: setVariableOpt(String, String),
setVariableOptEsc(String, String) and addBlockOpt(String).
EPL license added.

2009-01-22 chdh
Method getVariables() added.

2007-05-19 chdh
Conditional blocks ($if statement) implemented.
New nested class {@link MiniTemplator.TemplateSpecification}.
Old constructors replaced by the new general constructor
MiniTemplator(TemplateSpecification).
Parameter type of generateOutput(String) changed form File to String.
Second MiniTemplatorCache constructor removed.
Parameters of MiniTemplatorCache.get() method changed.

2006-10-18 chdh
New variant of addBlock() added with an isOptional parameter.

2006-07-07 chdh
Extended MiniTemplator and MiniTemplatorCache constructors with
charset argument added.

2004-11-20 chdh
"$Include" command changed so that the command text is not copied to
the output file.

2004-11-07 chdh
Method MiniTemplatorCache.clear() added.

2004-11-06 chdh
Changes for Java 5. (Unfortunately this version of MiniTemplator is no
longer compatible with Java 1.4).
"$Include" command implemented. Method loadSubtemplate and a new
constructor variant added.
Method cloneReset and class MiniTemplatorCache added.
MiniTemplatorCache Module created.

2004-10-28 chdh
Multiple blocks with the same name may now occur within a template.
No syntax error exception ("unknown command") is thrown any more, if a
HTML comment starts with "${".
serialVersionUID added to exception classes (for Java 5 compatibility).

2004-04-19 chdh
Methods blockExists, setVariableEsc and escapeHtml added.

2004-04-07 chdh
Parameter isOptional added to method setVariable.
Licensing changed from GPL to LGPL.

2003-07-16 chdh
Method setVariable() changed to throw an exception when the variable
does not exist (instead of returning false).

2003-07-08 chdh
Method variableExists() added.

2003-03-25 chdh
MiniTemplator converted from Visual Basic to Java.

2001-10-24 chdh (Christian d'Heureuse)
Visual Basic version of MiniTemplator created.

src/test/java/examples/Example1.java
// A simple example of how to use the MiniTemplator class.
package examples;

import biz.source_code.miniTemplator.MiniTemplator;
import java.nio.charset.Charset;

public class Example1 {

private static final String templateFileName = "src/test/resources/example1-template.html";
private static final String outputFileName   = "target/example1-output.html";

public static void main (String args[]) throws Exception {
    
   java.io.File f = new java.io.File(templateFileName);
   
   MiniTemplator t = new MiniTemplator.Builder()
                            .build(f.toURI().toURL(), Charset.defaultCharset());
   t.setVariable("animal1", "fox");
   t.setVariable("animal2", "dog");
   t.addBlock("block1");
   t.setVariable("animal1", "horse");
   t.setVariable("animal2", "cow");
   t.addBlock("block1");
   t.generateOutput(outputFileName); }

} // end class Example1

src/test/java/examples/Example2.java
// A more complex example of how to use the MiniTemplator class.
package examples;

import biz.source_code.miniTemplator.MiniTemplator;
import java.nio.charset.Charset;


public class Example2 {

private static final String templateFileName = "src/test/resources/example2-template.html";
private static final String outputFileName   = "target/example2-output.html";

public static void main (String args[]) throws Exception {
   java.io.File f = new java.io.File(templateFileName);
 
   MiniTemplator t = new MiniTemplator.Builder()
                    .build(f.toURI().toURL(), Charset.defaultCharset());
   t.setVariable("year", "2003");
   t.setVariable("month", "April");
   for (int weekOfYear=14; weekOfYear<=18; weekOfYear++) {
      for (int dayOfWeek=0; dayOfWeek<7; dayOfWeek++) {
         int dayOfMonth = (weekOfYear*7 + dayOfWeek) - 98;
         if (dayOfMonth >= 1 && dayOfMonth <= 30)
            t.setVariable("dayOfMonth", Integer.toString(dayOfMonth));
          else
            t.setVariable("dayOfMonth", "&nbsp;");
         t.addBlock("day"); }
      t.setVariable("weekOfYear", Integer.toString(weekOfYear));
      t.addBlock("week"); }
   t.generateOutput(outputFileName); }

} // end class Example2

src/examples/Example2.java
// A more complex example of how to use the MiniTemplator class.

import biz.source_code.miniTemplator.MiniTemplator;

public class Example2 {

private static final String templateFileName = "src/examples/example2-template.html";
private static final String outputFileName   = "target/examples-output/example2-output.html";

public static void main (String args[]) throws Exception {
   MiniTemplator t = new MiniTemplator(templateFileName);
   t.setVariable("year", "2003");
   t.setVariable("month", "April");
   for (int weekOfYear=14; weekOfYear<=18; weekOfYear++) {
      for (int dayOfWeek=0; dayOfWeek<7; dayOfWeek++) {
         int dayOfMonth = (weekOfYear*7 + dayOfWeek) - 98;
         if (dayOfMonth >= 1 && dayOfMonth <= 30)
            t.setVariable("dayOfMonth", Integer.toString(dayOfMonth));
          else
            t.setVariable("dayOfMonth", "&nbsp;");
         t.addBlock("day"); }
      t.setVariable("weekOfYear", Integer.toString(weekOfYear));
      t.addBlock("week"); }
   t.generateOutput(outputFileName); }

} // end class Example2

src/main/java/biz/source_code/miniTemplator/MiniTemplatorCache.java
// Copyright 2003-2010 Christian d'Heureuse, Inventec Informatik AG, Zurich, Switzerland
// www.source-code.biz, www.inventec.ch/chdh
//
// This module is multi-licensed and may be used under the terms
// of any of the following licenses:
//
//  EPL, Eclipse Public License, http://www.eclipse.org/legal
//  LGPL, GNU Lesser General Public License, http://www.gnu.org/licenses/lgpl.html
//
// This module is provided "as is", without warranties of any kind.

package biz.source_code.miniTemplator;

import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.util.HashMap;

/**
* A cache manager for MiniTemplator objects.
* This class is used to cache MiniTemplator objects in memory, so that
* each template file is only read and parsed once.
*
* <p>
* Home page: <a href="http://www.source-code.biz/MiniTemplator">www.source-code.biz/MiniTemplator</a><br>
* Author: Christian d'Heureuse, Inventec Informatik AG, Zurich, Switzerland<br>
* Multi-licensed: EPL/LGPL.
*/
public class MiniTemplatorCache {

private HashMap<String,MiniTemplator> cache;               // buffered MiniTemplator objects

/**
* Creates a new MiniTemplatorCache object.
*/
public MiniTemplatorCache() {
   cache = new HashMap<String,MiniTemplator>(); }

/**
* Returns a cloned MiniTemplator object from the cache.
* If there is not yet a MiniTemplator object with the specified <code>templateFileName</code>
* in the cache, a new MiniTemplator object is created and stored in the cache.
* Then the cached MiniTemplator object is cloned and the clone object is returned.
* @param  templateSpec      the template specification.
* @return                   a cloned and reset MiniTemplator object.
*/
public synchronized MiniTemplator get (MiniTemplator.TemplateSpecification templateSpec)
      throws IOException, MiniTemplator.TemplateSyntaxException {
   String key = generateCacheKey(templateSpec);
   MiniTemplator mt = cache.get(key);
   if (mt == null) {
      mt = new MiniTemplator(templateSpec);
      cache.put(key, mt); }
   return mt.cloneReset(); }

private static String generateCacheKey (MiniTemplator.TemplateSpecification templateSpec) {
   StringBuilder key = new StringBuilder(128);
   if (templateSpec.templateText != null)
      key.append(templateSpec.templateText);
    else if (templateSpec.templateFileName != null)
      key.append(templateSpec.templateFileName);
    else
      throw new IllegalArgumentException("No templateFileName or templateText specified.");
   if (templateSpec.conditionFlags != null) {
      for (String flag : templateSpec.conditionFlags) {
         key.append('|');
         key.append(flag.toUpperCase()); }}
   return key.toString(); }

/**
* Clears the cache.
*/
public synchronized void clear() {
   cache.clear(); }

} // end class MiniTemplatorCache

src/samples/Sample2.java
// A more complex example of how to use the MiniTemplator class.

import biz.source_code.miniTemplator.MiniTemplator;

public class Sample2 {

private static final String templateFileName = "src/samples/sample2-template.htm";
private static final String outputFileName   = "target/samples-output/sample2-output.htm";

public static void main (String args[]) throws Exception {
   MiniTemplator t = new MiniTemplator(templateFileName);
   t.setVariable ("year", "2003");
   t.setVariable ("month", "April");
   for (int weekOfYear=14; weekOfYear<=18; weekOfYear++) {
      for (int dayOfWeek=0; dayOfWeek<7; dayOfWeek++) {
         int dayOfMonth = (weekOfYear*7 + dayOfWeek) - 98;
         if (dayOfMonth >= 1 && dayOfMonth <= 30)
            t.setVariable ("dayOfMonth", Integer.toString(dayOfMonth));
          else
            t.setVariable ("dayOfMonth", "&nbsp;");
         t.addBlock ("day"); }
      t.setVariable ("weekOfYear", Integer.toString(weekOfYear));
      t.addBlock ("week"); }
   t.generateOutput (outputFileName); }

} // end class Sample2
